README.md
# A801TranslationFetcher
Fetches translation files for Atelier 801 games, decompresses them, then displays them.

Big thanks to [Alexunutrei](http://atelier801.com/profile?pr=Alexunutrei) for helping me understand the file encryption and how to parse the data.


CHANGELOG
== V1.6 - 20 January 2018 ==
- Added Transformice Adventures translation file option


== V1.5 - 15 February 2018 ==
- Game buttons now default to English, with language selection buttons below them.


== V1.4 - 27 January 2018 ==
- Share links now user URL variables instead of hash string
- Translation file is now cached on the server to save time fetching it.
- Added functionality for searching columns for specific text.


== V1.3 - 2 December 2017 ==
- Re-added project list to header and fixed it
- Added "permalinks". if the page is loaded with a permalink hash, it will skip directly to loading that file and then jump to the value.
- To avoid load errors, if translation file can't be fetched the php file will try 2 more times.


== V1.2 - 4 July 2017 ==
- Added syntax highlighting for "_P_" and "#SPEAKER_NAME:" Dead Maze syntax


== V1.1 - 4 June 2017 ==
- Now properly displays all characters.
- Fleshed out side design a bit more
- Certain translation syntax is now highlighted: %1 type variables, (male|female) syntax, and tags.
- Fixed script to properly show error message.


== V1.0 - 3 June 2017 ==
- Basic structure; scripts works from start to finish and can fetch and display file.


src/fetch_translation_file.php
<?php
# Return format
$format = "html";
if(isset($_GET["format"])) {
	switch($_GET["format"]) {
		case "json":
		case "text":
			$format = $_GET["format"];
			break;
	}
}

# Select the game
$game = "transformice";
if(isset($_GET["game"])) {
	switch($_GET["game"]) {
		case "transformice":
		case "tfm":
			$game = "transformice";
			break;
		case "deadmaze":
		case "dm":
			$game = "deadmaze";
			break;
		case "transformice_adventures":
		case "tfmadv":
			$game = "transformice_adventures";
			break;
		default:
			sendError(400, "No recognized game '{$_GET["game"]}'", $format);
			break;
	}
}

# Select the language
$language = isset($_GET["lang"]) ? clean($_GET["lang"]) : "en";

/********
* Build url
*********/
switch($game) {
	case "transformice":
		$filename = "tfz_$language";
		$url = "http://transformice.com/langues/$filename";
		break;
	case "deadmaze":
		$filename = "deadmeat_$language";
		$url = "http://transformice.com/langues/$filename";
		break;
	case "transformice_adventures":
		$filename = "tfmadv_$language";
		$url = "http://transformice.com/langues/$filename";
		break;
}

/********
* Check if cache data exists and is up-to-date
*********/
$local_file = "i18n/$filename.$format";
if(file_exists($local_file) && ($localDate = filemtime($local_file)) >= externalLastModified($url)) {
	// If cached file is good
	header('Last-Fetched-From-Game: '.gmdate('D, d M Y H:i:s ', $localDate) . 'GMT');
	$data = file_get_contents($local_file);
} else {
	// If file needs to be fetched
	/********
	* Get Data
	*********/
	try {
		$i = 3;
		do {
			if($i < 3) usleep(500000);
			$i--;
			$data = externalFetch($url);
			if($game != "transformice_adventures") {
				$data = gzuncompress($data);
			}
		} while (!$data && $i > 0);
		if(!$data) {
			sendError(400, "No lang file found for '$game': $language - (1) $filename", $format);
		}
	}
	catch (Exception $e) {
		sendError(400, "No lang file found for '$game': $language - (2) $filename", $format);
	}
	// $data = utf8_decode($data);
	// iconv("UTF-8", "CP1252", $data);
	// echo mb_detect_encoding($str, "auto")." --- ";
	// $data = htmlspecialchars($data);
	
	/********
	* Return Data
	*********/
	switch($format) {
		case "json":
			$data = charset_decode_utf_8($data);
			$split_char = "&#164;";
			// $split_char = "\xa4";
			$data = rtrim($data, $split_char);
			$data = explode($split_char, $data);
			// json_encode can't handle UTF-8 characters https://stackoverflow.com/q/6771938/1411473
			// mb_internal_encoding('UTF-8');
			// $data=array_map('utf8_encode',$data);
			
			// Convert array to json
			$data = json_encode($data);
			break;
		case "text":
			// Just use the default data
			// $data = charset_decode_utf_8($data);
			break;
		case "html":
		default:
			$data = charset_decode_utf_8($data);
			$data = "<html><body>$data</body></html>";
			break;
	}
	if(!is_dir("i18n")) { mkdir("i18n"); }
	file_put_contents($local_file, $data);
}

# Set appropriate headers
switch($format) {
	case "json":
		header('Content-type: application/json charset=utf-8');
		break;
	case "text":
		header('Content-Type: text/plain charset=utf-8');
		break;
	case "html":
		header('Content-Type: text/html charset=utf-8');
		break;
}
echo $data;

function externalFetch($url) {
	$ch = curl_init();
	curl_setopt($ch, CURLOPT_URL, $url);
	curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
	curl_setopt($ch, CURLOPT_SSLVERSION,3);
	$data = curl_exec ($ch);
	$error = curl_error($ch);
	curl_close ($ch);
	
	return $data;
}

function externalLastModified($url) {
	$h = get_headers($url, 1);
	if (!($h || strstr($h[0], '200') === false)) {
		return new \DateTime($h['Last-Modified']);//php 5.3
	}
	return null;
}

// http://php.net/manual/en/function.utf8-decode.php#116671
function charset_decode_utf_8 ($string) {
    /* Only do the slow convert if there are 8-bit characters */
    /* avoid using 0xA0 (\240) in ereg ranges. RH73 does not like that */
    if (!preg_match("/[\200-\237]/", $string)
     && !preg_match("/[\241-\377]/", $string)
    ) {
        return $string;
    }

    // decode three byte unicode characters
    $string = preg_replace("/([\340-\357])([\200-\277])([\200-\277])/e",
        "'&#'.((ord('\\1')-224)*4096 + (ord('\\2')-128)*64 + (ord('\\3')-128)).';'",
        $string
    );

    // decode two byte unicode characters
    $string = preg_replace("/([\300-\337])([\200-\277])/e",
        "'&#'.((ord('\\1')-192)*64+(ord('\\2')-128)).';'",
        $string
    );

    return $string;
}

function clean($str){
	$re = "/([0-9a-zA-Z])/u";
	preg_match_all($re, $str, $matches);
	return isset($matches[0]) ? implode($matches[0]) : '';
}

function sendError($code, $msg, $format) {
	header("HTTP/1.0 $code $msg");
	// switch($format) {
	// 	case "json":
	// 		header('Content-type: application/json');
	// 		echo json_encode(array( "error" => true, "error_msg" => $msg ));
	// 		break;
	// 	case "text":
	// 	default:
			echo $msg;
	// 		break;
	// }
	http_response_code($code);
	exit;
}


src/js/index.js
(function(){
	// Grabs the url vars
	var URL_VARS = {}; window.location.href.split("#")[0].replace( /[?&]+([^=&]+)=([^&]*)/gi, function(m,key,value){URL_VARS[key] = value;} );
	
	let game = null;//"tfm";
	let lang = null;//"en";
	let id = null;
	
	setActiveStep("game-step");
	
	if(URL_VARS.g) {
		game = URL_VARS.g;
		lang = URL_VARS.l;
		id = URL_VARS.id;
		if(lang && game) {
			doRetrieval();
		}
		else if(game) {
			setActiveStep("lang-step");
		}
	} else {
		// Backwards compatability for old hash links
		let startHash = window.location.hash.substr(1);
		if(startHash) {
			let [tGame, tLang, tID] = startHash.split(",");
			if(tID) {
				lang = tLang;
				game = tGame;
				id = tID;
				doRetrieval();
			}
		}
	}
	
	/**************
	* Event Listeners
	**************/
	addEventListenerAll(document.querySelectorAll(".js-game-chosen"), "click", onGameClicked);
	addEventListenerAll(document.querySelectorAll(".js-game-lang-chosen"), "click", onGameLangClicked);
	addEventListenerAll(document.querySelectorAll(".js-lang-chosen"), "click", onLangClicked);
	addEventListenerAll(document.querySelectorAll(".js-lang-chosen-custom"), "click", onCustomLangEntered);
	addEventListenerAll(document.querySelectorAll(".js-reset"), "click", onResetClicked);
	
	function onGameClicked(e) {
		setActiveStep("lang-step");
		game = e.target.dataset.game;
		updateUrl();
	}
	
	function onLangClicked(e) {
		lang = e.target.dataset.lang;
		updateUrl();
		doRetrieval();
	}
	
	function onGameLangClicked(e) {
		game = e.target.dataset.game;
		lang = e.target.dataset.lang;
		updateUrl();
		doRetrieval();
	}
	
	function onCustomLangEntered(e) {
		lang = document.querySelector("#lang_code").value;
		if(lang.length < 2) { return; }
		doRetrieval();
	}
	
	function doRetrieval() {
		setActiveStep("fetch-step");
		fetchI18nData(game, lang, (pData, xmlhttp) => {
			addDataToPage(pData);
			updateUrl();
			if(id) {
				// Reactivate hash to go to ID
				onPermaClick(id);
				// window.location.href = "#"+startHash;
			}
			
			if(xmlhttp.getResponseHeader("Last-Fetched-From-Game")) {
				document.querySelector("#result-timestamp").innerHTML = "Translation files last cached: "+xmlhttp.getResponseHeader("Last-Fetched-From-Game");
			} else {
				document.querySelector("#result-timestamp").innerHTML = "";
			}
		}, doDataRetreivalError);
	}
	
	function onResetClicked(e) {
		game = null;
		lang = null;
		id = null;
		updateUrl();
		setActiveStep("game-step");
	}
	
	function updateUrl() {
		history.replaceState({}, null, getShareUrl(game, lang, id));
	}
	function getShareUrl(game, lang, id) {
		var tParams = [];
		if(id) { tParams.push("id="+id); }
		if(game) { tParams.push("g="+game); }
		if(lang) { tParams.push("l="+lang); }
		return location.origin + location.pathname + (tParams.length > 0 ? "?"+tParams.join("&") : "");
	}
	
	/**************
	* Methods
	**************/
	function setActiveStep(pID) {
		let tSteps = document.querySelectorAll(".step");
		for(let i = 0; i < tSteps.length; i++) {
			tSteps[i].classList.remove("active");
		}
		// Doesn't work in Edge
		// for(tStep of tSteps) {
		// 	tStep.classList.remove("active");
		// }
		document.querySelector("#"+pID).classList.add("active");
	}
	
	function addDataToPage(pData) {
		if(!pData) {
			doDataRetreivalError("Server returned no content.");
		}
		setActiveStep("result-step");
		// let tLines = pData;
		pData = pData.replace(/¤$/, ''); // Replace extra one at end
		let tLines = pData.split("¤");
		// let tLines = pData.split("&#164;");
		let tHTML = "";
		tHTML += "<table class='result-table'>";
		// Add table head
		tHTML += "<thead>";
			tHTML += "<tr>";
			tHTML += "<th>Unique Key</th>";
			tHTML += "<th>Translation</th>";
			tHTML += "</tr>";
		tHTML += "</thead>";
		// Add table contents
		tHTML += "<tbody>";
		let tKey, tMessage, tHashTag;
		for(tLine of tLines) {
			[tKey, tMessage] = splitOnce(tLine, "=");
			tMessage = highlightSyntaxAll(tMessage);
			tMessage = `<pre>${tMessage}</pre>`;
			tHashTag = tKey;
			// Extra row before it is needed for some CSS styling
			tHTML += `<tr id="${tHashTag}" class="permalink-target"></tr>`
				+`<tr><th><a class='permalink' href="${getShareUrl(game, lang, tKey)}" onclick="onPermaClick('${tKey}'); return false;">#</a><div class="overflow">${tKey}</div></th><td>${tMessage}</td></tr>`;
		}
		tHTML += "</tbody>";
		tHTML += "</table>";
		document.querySelector("#result").innerHTML = tHTML;
		
		// Allow searching of table
		$(".result-table").tablesorter({
			widthFixed: true,
			removeRows: true,
			widgets : ['filter', 'columns'],
			widgetOptions: {
				filter_placeholder: { search:"Search column..." },
				filter_searchDelay : 500,
			}
		});
	}
	
	function onPermaClick(pKey) {
		id = pKey;
		window.location.href = "#"+pKey; // Jumps user to id
		updateUrl(); // Clears the hash
	}
	window.onPermaClick = onPermaClick;
	
	function doDataRetreivalError(pMessage) {
		setActiveStep("error-step");
		document.querySelector("#error").innerHTML = pMessage;
	}
	
	/**************
	* Helper Methods
	**************/
	function fetchI18nData(pGame, pLang, pCallback, pFailCallback) {
		let tUrl = `fetch_translation_file.php?game=${pGame}&lang=${pLang}&format=text`;
		// fetch(tUrl, {
		// 	method:"GET",
		// })
		// .then(function(response) { return response.text(); })
		// .then(function(text){
		// 	pCallback(text);
		// });
		fewAjax({
			url:tUrl, method:"GET", dataType:"text",
			success:pCallback, fail:pFailCallback,
		});
	}
	
	function addEventListenerAll(pList, pEventName, pCallback) {
		for(tElem of pList) {
			tElem.addEventListener(pEventName, pCallback);
		}
	}
	
	function highlightSyntaxAll(pString) {
		pString = highlightHTML(pString);
		pString = highlightStringSubstitution(pString);
		pString = highlightSex(pString);
		pString = highlightSpeaker(pString);
		pString = highlightDialogBreak(pString);
		return pString;
	}
	
	function highlightHTML(pString) {
		// let tTitle = htmlEscape(pString);
		// return `<span title='${tTitle}'>${highlightHTMLRecurrsive(pString)}</span>`;
		return `<!-- ${pString} -->${highlightHTMLRecurrsive(pString)}`
		// return pString.replace(/</g, "&lt;").replace(/>/g, "&gt;");
	}
	function highlightHTMLRecurrsive(pString) {
		let tRegex = /<(\w+)(?:\s*(.*?))>((?:.|\n)*?)((?:<\/\1>|$))/g;
		let tMatches = tRegex.exec(pString);
		if(tMatches) {
			let [tAllMatched, tTag, tAttribs, tContent, tEnd] = tMatches;
			// console.log(tMatches);
			tContent = highlightHTMLRecurrsive(tContent);
			let tFormattedCode;
			switch(tTag) {
				case "img":
					tAttribs = tAttribs.replace("http:", "https:");
					tFormattedCode = `<img ${tAttribs} title="${htmlEscape(tAttribs)}" />${tContent}`;
					break;
				case "font":
					if(tAttribs.indexOf("color=") > -1) {
						tContent = `<font ${tAttribs}>${tContent}</font>`;
					}
					// Fall through
				default:
					tAttribs = tAttribs ? ` <span class='attr'>${tAttribs}</span>` : "";
					tEnd = tEnd ? `<span class='tag'>&lt;/${tTag}&gt</span>` : "";
					tFormattedCode = `<span class='tag'>&lt;${tTag}${tAttribs}&gt;</span>${tContent}${tEnd}`;
					break;
			}
			pString = pString.replace(tAllMatched, tFormattedCode);
		}
		return pString;
	}
	
	function highlightStringSubstitution(pString) {
		return pString.replace(/(\%\d)/g, "<span class='subst'>$1</span>");
	}
	
	function highlightSex(pString) {
		return pString.replace(/\((.*?)\|(.*?)\)/g, "<span class='sex'>(<span class='m'>$1</span>|<span class='f'>$2</span>)</span>");
	}
	
	function highlightSpeaker(pString) {
		return pString.replace(/(^#[A-Za-z_]*:)/gm, "<span class='speaker'>$1</span>");
	}
	
	function highlightDialogBreak(pString) {
		return pString.replace(/(_P_)/g, "<span class='d-break'>$1</span>");
	}
	
	// pData = { url:String, method:String="GET", dataType:String="text", contentType?:String, success?:String->Void, fail?:String->Void }
	function fewAjax(pData) {
		pData.method = pData.method || "GET";
		pData.dataType = pData.dataType || "text";
		
		let xmlhttp = new XMLHttpRequest();

		xmlhttp.onreadystatechange = function() {
			if (xmlhttp.readyState == XMLHttpRequest.DONE ) {
				if (xmlhttp.status == 200) {
					if(pData.success) {
						let tResponse = xmlhttp.responseText;
						try {
							switch(pData.dataType) {
								case "json": tResponse = JSON.parse(tResponse); break;
							}
						} catch(e) {
							if(pData.fail) { pData.fail("Error parsing data as "+pData.dataType); }
							return;
						}
						pData.success(tResponse, xmlhttp);
					}
				}
				else {
					if(pData.fail) { pData.fail("Error "+xmlhttp.status+": "+xmlhttp.responseText); }
				}
			}
		};

		xmlhttp.open(pData.method, pData.url, true);
		if(pData.contentType) xmlhttp.setRequestHeader('Content-Type', pData.contentType);
		xmlhttp.send();
	}
	
	function htmlEscape(pString) {
		return pString.replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&#34;").replace(/'/g, "&#39;");
	}
	
	// Only split first instance of something in a string. ex: "a-b-c-d" split on "-" would return "a" and "b-c-d"
	function splitOnce(pString, pSeparator) {
		let tSplitIndex = pString.indexOf(pSeparator);
		return [ pString.substring(0, tSplitIndex), pString.substring(tSplitIndex+pSeparator.length) ];
	}
})();
